test/governance/extensions/GovernorStorage.test.js

```diff
@@ -0,0 +1,150 @@
+const { constants, expectEvent, expectRevert } = require('@openzeppelin/test-helpers');
+const { expect } = require('chai');
+
+const { expectRevertCustomError } = require('../../helpers/customError');
+const Enums = require('../../helpers/enums');
+const { GovernorHelper, timelockSalt } = require('../../helpers/governance');
+
+const Timelock = artifacts.require('TimelockController');
+const Governor = artifacts.require('$GovernorStorageMock');
+const CallReceiver = artifacts.require('CallReceiverMock');
+
+const TOKENS = [
+  { Token: artifacts.require('$ERC20Votes'), mode: 'blocknumber' },
+  { Token: artifacts.require('$ERC20VotesTimestampMock'), mode: 'timestamp' },
+];
+
+contract('GovernorStorage', function (accounts) {
+  const [owner, voter1, voter2, voter3, voter4] = accounts;
+
+  const DEFAULT_ADMIN_ROLE = '0x0000000000000000000000000000000000000000000000000000000000000000';
+  const PROPOSER_ROLE = web3.utils.soliditySha3('PROPOSER_ROLE');
+  const EXECUTOR_ROLE = web3.utils.soliditySha3('EXECUTOR_ROLE');
+  const CANCELLER_ROLE = web3.utils.soliditySha3('CANCELLER_ROLE');
+
+  const name = 'OZ-Governor';
+  const version = '1';
+  const tokenName = 'MockToken';
+  const tokenSymbol = 'MTKN';
+  const tokenSupply = web3.utils.toWei('100');
+  const votingDelay = web3.utils.toBN(4);
+  const votingPeriod = web3.utils.toBN(16);
+  const value = web3.utils.toWei('1');
+
+  for (const { mode, Token } of TOKENS) {
+    describe(`using ${Token._json.contractName}`, function () {
+      beforeEach(async function () {
+        const [deployer] = await web3.eth.getAccounts();
+
+        this.token = await Token.new(tokenName, tokenSymbol, tokenName, version);
+        this.timelock = await Timelock.new(3600, [], [], deployer);
+        this.mock = await Governor.new(
+          name,
+          votingDelay,
+          votingPeriod,
+          0,
+          this.timelock.address,
+          this.token.address,
+          0,
+        );
+        this.receiver = await CallReceiver.new();
+
+        this.helper = new GovernorHelper(this.mock, mode);
+
+        await web3.eth.sendTransaction({ from: owner, to: this.timelock.address, value });
+
+        // normal setup: governor is proposer, everyone is executor, timelock is its own admin
+        await this.timelock.grantRole(PROPOSER_ROLE, this.mock.address);
+        await this.timelock.grantRole(PROPOSER_ROLE, owner);
+        await this.timelock.grantRole(CANCELLER_ROLE, this.mock.address);
+        await this.timelock.grantRole(CANCELLER_ROLE, owner);
+        await this.timelock.grantRole(EXECUTOR_ROLE, constants.ZERO_ADDRESS);
+        await this.timelock.revokeRole(DEFAULT_ADMIN_ROLE, deployer);
```

Comment:
I tried `Promise.all` and regardless of `await` Hardhat will mine a block for every transaction. It doesn't seem to make a difference.

Reply:
I made a quick test and it was saving a considerable amount of time if instead of `Promise.all` we disabled Hardhat's automining. I created an util for it.

Take a look to https://github.com/OpenZeppelin/openzeppelin-contracts/pull/4360/commits/b2445f8bd86e49ea8f87769f9337f5ff641ecb16

I only applied it to GovernorStorage because I don't want to add even more changes to this PR, but these were the setup time reductions:

| Test Case                    | With Batching (ms) | Without Batching (ms) |
|------------------------------|--------------------|-----------------------|
| queue and execute by id      | 5.987              | 108.955               |
| cancel by id, proposal indexing | 5.412            | 93.074                |
| before propose               | 5.306              | 92.509                |
| after propose                | 1.833              | 80.512                |
| queue and execute by id      | 1.429              | 74.976                |
| cancel by id, proposal indexing | 1.434            | 81.671                |
| before propose               | 1.343              | 90.931                |
| after propose                | 1.297              | 83.321                |

Reply:
Seem very good, but we 100% need the `await` in each statement. The benefit probably comes from automine being off.

Reply:
It fails because each `await` is waiting for a block to be mined apparently. Shall we revert?

Reply:
Ah, I forgot Truffle did that. Let me see.

Reply:
We should just use the existing `batchInBlock`.

Reply:
It's not working, I don't understand. Let's revert.

Reply:
We resolved this problem a while ago, when working on the voting contract. We have `batchInBlock` for that:
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/test/helpers/txpool.js#L10-L33

Is that not nice enough for what you want to do ?

Reply:
Note that I don't get why you want all that to happen in a single block. We have similar setups in other test files, do you also want to change them ?

Reply:
The motivation was for performance reasons, allegedly tests run faster.

We should do this in another PR anyway.

Reply:
Yes, let's do it in another PR, I don't want to shift the attention to this.